# Práctica 3: Aprendizaje Automático
### 21722 - Inteligencia Artificial
#### Por: Andrea Valentina Reyes Gómez


Se quiere predecir el valor monetario de un jugador de fútbol utilizando la información del mismo.

## Importar librerías

In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler

import pandas as pd
import numpy as np

## Análisis de los datos

Como primer paso visualizamos los datos para saber qué debemos cambiar o mantener para que los datos permitan hacer predicciones lo más acertadas posible.

In [2]:
df = pd.read_csv(os.path.join("in", "fifa.csv"))
pd.set_option('display.max_columns', None)
df

Unnamed: 0      ID                Name  Age  \
0               0  158023            L. Messi   31   
1               1   20801   Cristiano Ronaldo   33   
2               2  190871           Neymar Jr   26   
3               3  193080              De Gea   27   
4               4  192985        K. De Bruyne   27   
...           ...     ...                 ...  ...   
18202       18202  238813        J. Lundstram   19   
18203       18203  243165  N. Christoffersson   19   
18204       18204  241638           B. Worman   16   
18205       18205  246268      D. Walker-Rice   17   
18206       18206  246269           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right    

Al mirar los datos de las columnas nos damos cuenta de que muchos de ellos son inválidos o irrelevantes en la decisión final del valor monetario de un jugador. Por lo tanto, procedemos a identificar y eliminar dichas columnas.

Las columnas **Unnamed**, **ID**, **Photo**, **Flag**, **Club Logo**, **Special**, **Preferred Foot**, **Real Face**, **Jersey Number**, **Joined** y **Contract Valid Until** son las que no proporcionan información relevante, puesto que no son datos medibles ni le suman o restan valor a un jugador. 

In [3]:
df.pop('Unnamed: 0')
df.pop('ID')
df.pop('Photo')
df.pop('Flag')
df.pop('Club Logo')
df.pop('Special')
df.pop("Preferred Foot")
df.pop('Real Face')
df.pop('Jersey Number')
df.pop('Joined')
df.pop('Contract Valid Until')

0        2021
1        2022
2        2022
3        2020
4        2023
         ... 
18202    2019
18203    2020
18204    2021
18205    2019
18206    2019
Name: Contract Valid Until, Length: 18207, dtype: object

Ahora observamos la tabla luego de eliminar algunos datos irrelevantes. 

In [4]:
df

Name  Age Nationality  Overall  Potential  \
0                L. Messi   31   Argentina       94         94   
1       Cristiano Ronaldo   33    Portugal       94         94   
2               Neymar Jr   26      Brazil       92         93   
3                  De Gea   27       Spain       91         93   
4            K. De Bruyne   27     Belgium       91         92   
...                   ...  ...         ...      ...        ...   
18202        J. Lundstram   19     England       47         65   
18203  N. Christoffersson   19      Sweden       47         63   
18204           B. Worman   16     England       47         67   
18205      D. Walker-Rice   17     England       47         66   
18206           G. Nugent   16     England       46         66   

                      Club    Value   Wage  International Reputation  \
0             FC Barcelona  €110.5M  €565K                       5.0   
1                 Juventus     €77M  €405K                       5.0   
2      Paris Saint-Germain  €118.5M  €290K                       5.0   
3        Manchester United     €72M  €260K                       4.0   
4          Manchester City    €102M  €355K                       4.0   
...                    ...      ...    ...                       ...   
18202      Crewe Alexandra     €60K    €1K                       1.0   
18203       Trelleborgs FF     €60K    €1K                       1.0   
18204     Cambridge United     €60K    €1K                       1.0   
18205      Tranmere Rovers     €60K    €1K                       1.0   
18206      Tranmere Rovers     €60K    €1K                       1.0   

       Weak Foot  Skill Moves       Work Rate   Body Type Position  \
0            4.0          4.0  Medium/ Medium       Messi       RF   
1            4.0          5.0       High/ Low  C. Ronaldo       ST   
2            5.0          5.0    High/ Medium      Neymar       LW   
3            3.0          1.0  Medium/ Medium        Lean       GK   
4            5.0          4.0      High/ High      Normal      RCM   
...          ...          ...             ...         ...      ...   
18202        2.0          2.0  Medium/ Medium        Lean       CM   
18203        2.0          2.0  Medium/ Medium      Normal       ST   
18204        3.0          2.0  Medium/ Medium      Normal       ST   
18205        3.0          2.0  Medium/ Medium        Lean       RW   
18206        3.0          2.0  Medium/ Medium        Lean       CM   

      Loaned From Height  Weight    LS    ST    RS    LW    LF    CF    RF  \
0             NaN    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2  93+2   
1             NaN    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3  90+3   
2             NaN    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3  89+3   
3             NaN    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4             NaN   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3  87+3   
...           ...    ...     ...   ...   ...   ...   ...   ...   ...   ...   
18202         NaN    5'9  134lbs  42+2  42+2  42+2  44+2  44+2  44+2  44+2   
18203         NaN    6'3  170lbs  45+2  45+2  45+2  39+2  42+2  42+2  42+2   
18204         NaN    5'8  148lbs  45+2  45+2  45+2  45+2  46+2  46+2  46+2   
18205         NaN   5'10  154lbs  47+2  47+2  47+2  47+2  46+2  46+2  46+2   
18206         NaN   5'10  176lbs  43+2  43+2  43+2  45+2  44+2  44+2  44+2   

         RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM  \
0      92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2   
1      89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3   
2      89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  44+2  45+2 

## Tratamiento de valores no numéricos

Podemos observar que hay ciertos datos que corresponden a valores monetarios, pero que en sus instancias incluyen el símbolo de la moneda en la que está expresado dicho valor. Como todos estos valores están expresados en euros, podemos remover el símbolo y así tener solo datos numéricos que nos permitan hacer mejores estimaciones.

Además, sabemos que K indica que el valor está en el rango de los miles y que M indica que está en el rango de los millones, por lo que también podemos eliminar estas letras y colocar su valor correspondiente.

Esto se realizará utilizando el código de transformación proporcionado en el enunciado de la práctica.

In [5]:
def value_to_float(x):
    """
    From K and M to float.
    Delete '€'
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Aplicamos esta función a cada valor de la columna **Value**.

In [6]:
df["Value"] = df["Value"].apply(value_to_float)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value   Wage  International Reputation  \
0         FC Barcelona  110500000.0  €565K                       5.0   
1             Juventus   77000000.0  €405K                       5.0   
2  Paris Saint-Germain  118500000.0  €290K                       5.0   
3    Manchester United   72000000.0  €260K                       4.0   
4      Manchester City  102000000.0  €355K                       4.0   

   Weak Foot  Skill Moves       Work Rate   Body Type Position Loaned From  \
0        4.0          4.0  Medium/ Medium       Messi       RF         NaN   
1        4.0          5.0       High/ Low  C. Ronaldo       ST         NaN   
2        5.0          5.0    High/ Medium      Neymar       LW         NaN   
3        3.0          1.0  Medium/ Medium        Lean       GK         NaN   
4        5.0          4.0      High/ High      Normal      RCM         NaN   

  Height  Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   

    RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   

    LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0      75.0       91.0  

Aplicamos la misma función al resto de columnas que contienen valores monetarios: **Wage** y **Release Clause**.

In [7]:
df["Wage"] = df["Wage"].apply(value_to_float)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  International Reputation  \
0         FC Barcelona  110500000.0  565000.0                       5.0   
1             Juventus   77000000.0  405000.0                       5.0   
2  Paris Saint-Germain  118500000.0  290000.0                       5.0   
3    Manchester United   72000000.0  260000.0                       4.0   
4      Manchester City  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves       Work Rate   Body Type Position Loaned From  \
0        4.0          4.0  Medium/ Medium       Messi       RF         NaN   
1        4.0          5.0       High/ Low  C. Ronaldo       ST         NaN   
2        5.0          5.0    High/ Medium      Neymar       LW         NaN   
3        3.0          1.0  Medium/ Medium        Lean       GK         NaN   
4        5.0          4.0      High/ High      Normal      RCM         NaN   

  Height  Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   

    RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   

    LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0     

Cambiamos el tipo de datos de la columna **Release Clause** a *str* y aplicamos la función de transformación definida anteriormente.

In [8]:
df["Release Clause"] = df["Release Clause"].astype(str)

In [9]:
df["Release Clause"] = df["Release Clause"].apply(value_to_float)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  International Reputation  \
0         FC Barcelona  110500000.0  565000.0                       5.0   
1             Juventus   77000000.0  405000.0                       5.0   
2  Paris Saint-Germain  118500000.0  290000.0                       5.0   
3    Manchester United   72000000.0  260000.0                       4.0   
4      Manchester City  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves       Work Rate   Body Type Position Loaned From  \
0        4.0          4.0  Medium/ Medium       Messi       RF         NaN   
1        4.0          5.0       High/ Low  C. Ronaldo       ST         NaN   
2        5.0          5.0    High/ Medium      Neymar       LW         NaN   
3        3.0          1.0  Medium/ Medium        Lean       GK         NaN   
4        5.0          4.0      High/ High      Normal      RCM         NaN   

  Height  Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   

    RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   

    LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0     

Vemos que en la columna **Weight** los valores tienen como sufijo *'lbs'* para indicar que se mide en libras. Como la unidad es la misma para todos los pesos, eliminamos el sufijo para quedarnos solo con el valor numérico.

In [10]:
def replace_lbs(x):
    """
    Delete 'lbs' from weight
    
    """
    x = x.replace('lbs', '')
    
    return x

In [11]:
df["Weight"] = df["Weight"].astype(str)

In [12]:
df["Weight"] = df["Weight"].apply(replace_lbs)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  International Reputation  \
0         FC Barcelona  110500000.0  565000.0                       5.0   
1             Juventus   77000000.0  405000.0                       5.0   
2  Paris Saint-Germain  118500000.0  290000.0                       5.0   
3    Manchester United   72000000.0  260000.0                       4.0   
4      Manchester City  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves       Work Rate   Body Type Position Loaned From  \
0        4.0          4.0  Medium/ Medium       Messi       RF         NaN   
1        4.0          5.0       High/ Low  C. Ronaldo       ST         NaN   
2        5.0          5.0    High/ Medium      Neymar       LW         NaN   
3        3.0          1.0  Medium/ Medium        Lean       GK         NaN   
4        5.0          4.0      High/ High      Normal      RCM         NaN   

  Height Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0    5'7    159  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1    6'2    183  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2    5'9    150  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3    6'4    168   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   5'11    154  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   

    RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   

    LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0      75.0 

Por su parte, la columna **Height** tiene valores con un apóstrofe en lugar de un punto. Lo reemplazaremos para convertir este valor en numérico.

In [13]:
def replace_apostrophe(x):
    """
    Delete apostrophe from height
    
    """
    x = x.replace('\'', '.')
    
    return x

In [14]:
df["Height"] = df["Height"].astype(str)

In [15]:
df["Height"] = df["Height"].apply(replace_apostrophe)
df.head()

Name  Age Nationality  Overall  Potential  \
0           L. Messi   31   Argentina       94         94   
1  Cristiano Ronaldo   33    Portugal       94         94   
2          Neymar Jr   26      Brazil       92         93   
3             De Gea   27       Spain       91         93   
4       K. De Bruyne   27     Belgium       91         92   

                  Club        Value      Wage  International Reputation  \
0         FC Barcelona  110500000.0  565000.0                       5.0   
1             Juventus   77000000.0  405000.0                       5.0   
2  Paris Saint-Germain  118500000.0  290000.0                       5.0   
3    Manchester United   72000000.0  260000.0                       4.0   
4      Manchester City  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves       Work Rate   Body Type Position Loaned From  \
0        4.0          4.0  Medium/ Medium       Messi       RF         NaN   
1        4.0          5.0       High/ Low  C. Ronaldo       ST         NaN   
2        5.0          5.0    High/ Medium      Neymar       LW         NaN   
3        3.0          1.0  Medium/ Medium        Lean       GK         NaN   
4        5.0          4.0      High/ High      Normal      RCM         NaN   

  Height Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0    5.7    159  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1    6.2    183  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2    5.9    150  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3    6.4    168   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   5.11    154  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   

    RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   

    LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0      75.0 

Otra característica que influye en el precio de un futbolista es su **nacionalidad**, ya que para ciertas competiciones se limita a los clubes en cuando a la cantidad de jugadores extranjeros, etc. Sin embargo, los valores de esta característica no están presentados de forma numérica en este set de datos, por lo cual transformaremos esta columna en varias, una para cada nacionalidad y colocaremos un 1 si el jugador es de esa nacionalidad o un 0 en caso contrario.

In [16]:
nationality = df.pop("Nationality")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nationality, prefix='Nationality').reset_index(drop=True)], axis=1, sort=False)
df.head()

Name  Age  Overall  Potential                 Club  \
0           L. Messi   31       94         94         FC Barcelona   
1  Cristiano Ronaldo   33       94         94             Juventus   
2          Neymar Jr   26       92         93  Paris Saint-Germain   
3             De Gea   27       91         93    Manchester United   
4       K. De Bruyne   27       91         92      Manchester City   

         Value      Wage  International Reputation  Weak Foot  Skill Moves  \
0  110500000.0  565000.0                       5.0        4.0          4.0   
1   77000000.0  405000.0                       5.0        4.0          5.0   
2  118500000.0  290000.0                       5.0        5.0          5.0   
3   72000000.0  260000.0                       4.0        3.0          1.0   
4  102000000.0  355000.0                       4.0        5.0          4.0   

        Work Rate   Body Type Position Loaned From Height Weight    LS    ST  \
0  Medium/ Medium       Messi       RF         NaN    5.7    159  88+2  88+2   
1       High/ Low  C. Ronaldo       ST         NaN    6.2    183  91+3  91+3   
2    High/ Medium      Neymar       LW         NaN    5.9    150  84+3  84+3   
3  Medium/ Medium        Lean       GK         NaN    6.4    168   NaN   NaN   
4      High/ High      Normal      RCM         NaN   5.11    154  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0  

## Tratamiento de valores NaN

Ahora nos dedicamos a tratar aquellos datos que tienen valores inválidos. Para ello, primero buscamos a qué columnas pertenecen esos valores

In [17]:
df.columns[df.isna().any()].tolist()

['Club',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Body Type',
 'Position',
 'Loaned From',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes']

In [18]:
df[df.isna().any(axis=1)]

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

             Value      Wage  International Reputation  Weak Foot  \
0      110500000.0  565000.0                       5.0        4.0   
1       77000000.0  405000.0                       5.0        4.0   
2      118500000.0  290000.0                       5.0        5.0   
3       72000000.0  260000.0                       4.0        3.0   
4      102000000.0  355000.0                       4.0        5.0   
...            ...       ...                       ...        ...   
18202      60000.0    1000.0                       1.0        2.0   
18203      60000.0    1000.0                       1.0        2.0   
18204      60000.0    1000.0                       1.0        3.0   
18205      60000.0    1000.0                       1.0        3.0   
18206      60000.0    1000.0                       1.0        3.0   

       Skill Moves       Work Rate   Body Type Position Loaned From Height  \
0              4.0  Medium/ Medium       Messi       RF         NaN    5.7   
1              5.0       High/ Low  C. Ronaldo       ST         NaN    6.2   
2              5.0    High/ Medium      Neymar       LW         NaN    5.9   
3              1.0  Medium/ Medium        Lean       GK         NaN    6.4   
4              4.0      High/ High      Normal      RCM         NaN   5.11   
...            ...             ...         ...      ...         ...    ...   
18202          2.0  Medium/ Medium        Lean       CM         NaN    5.9   
18203          2.0  Medium/ Medium      Normal       ST         NaN    6.3   
18204          2.0  Medium/ Medium      Normal       ST         NaN    5.8   
18205          2.0  Medium/ Medium        Lean       RW         NaN   5.10   
18206          2.0  Medium/ Medium        Lean       CM         NaN   5.10   

      Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0        159  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1        183  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2        150  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3        168   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4        154  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   
...      ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202    134  42+2  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2   
18203    170  45+2  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2   
18204    148  45+2  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2   
18205    154  47+2  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2   
18206    176  43+2  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0      93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1      88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2      89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      88+3  88+3  87+3  87+3  87+3 

De aquí podemos observar que la mayoría de las columnas tienen al menos un valor NaN (Not a Number). 

Para tratar estos datos podemos empezar por eliminar la columna **Loaned From**, ya que no aporta información relevante.

In [19]:
df.pop("Loaned From")
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

             Value      Wage  International Reputation  Weak Foot  \
0      110500000.0  565000.0                       5.0        4.0   
1       77000000.0  405000.0                       5.0        4.0   
2      118500000.0  290000.0                       5.0        5.0   
3       72000000.0  260000.0                       4.0        3.0   
4      102000000.0  355000.0                       4.0        5.0   
...            ...       ...                       ...        ...   
18202      60000.0    1000.0                       1.0        2.0   
18203      60000.0    1000.0                       1.0        2.0   
18204      60000.0    1000.0                       1.0        3.0   
18205      60000.0    1000.0                       1.0        3.0   
18206      60000.0    1000.0                       1.0        3.0   

       Skill Moves       Work Rate   Body Type Position Height Weight    LS  \
0              4.0  Medium/ Medium       Messi       RF    5.7    159  88+2   
1              5.0       High/ Low  C. Ronaldo       ST    6.2    183  91+3   
2              5.0    High/ Medium      Neymar       LW    5.9    150  84+3   
3              1.0  Medium/ Medium        Lean       GK    6.4    168   NaN   
4              4.0      High/ High      Normal      RCM   5.11    154  82+3   
...            ...             ...         ...      ...    ...    ...   ...   
18202          2.0  Medium/ Medium        Lean       CM    5.9    134  42+2   
18203          2.0  Medium/ Medium      Normal       ST    6.3    170  45+2   
18204          2.0  Medium/ Medium      Normal       ST    5.8    148  45+2   
18205          2.0  Medium/ Medium        Lean       RW   5.10    154  47+2   
18206          2.0  Medium/ Medium        Lean       CM   5.10    176  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
18204  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
18205  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
18206  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   N

Ahora, analizamos para cada columna aquellas casillas que tengan valores NaN y veremos cómo tratar cada una de ellas. Empezamos con el atributo **International Reputation**

In [20]:
df.loc[df["International Reputation"] != df["International Reputation"]]

Name  Age  Overall  Potential                           Club  \
13236      J. McNulty   33       62         62                       Rochdale   
13237      J. Barrera   29       62         62                Boyacá Chicó FC   
13238        J. Stead   35       62         62                   Notts County   
13239     A. Semprini   20       62         72                        Brescia   
13240      R. Bingham   24       62         66         Hamilton Academical FC   
13241    K. Dankowski   21       62         72                  Śląsk Wrocław   
13242       I. Colman   23       62         70         Club Atlético Aldosivi   
13243       M. Feeney   19       62         78                        Everton   
13244        R. Minor   30       62         62                       Hobro IK   
13245          Klauss   21       62         69                   HJK Helsinki   
13246      I. Sissoko   22       62         68                     AS Béziers   
13247         F. Hart   28       62         62                 SV Mattersburg   
13248   L. McCullough   24       62         69                Tranmere Rovers   
13249       Li Yunqiu   27       62         62  Shanghai Greenland Shenhua FC   
13250       F. Garcia   29       62         62               Itagüí Leones FC   
13251    R. Haemhouts   34       62         62                      NAC Breda   
13252       E. Binaku   22       62         70                       Malmö FF   
13253       G. Miller   31       62         62                Carlisle United   
13254      A. Aidonis   17       62         82                  VfB Stuttgart   
13255        L. Sowah   25       62         65         Hamilton Academical FC   
13256       R. Deacon   26       62         62                      Dundee FC   
13257   Jang Hyun Soo   25       62         65        Suwon Samsung Bluewings   
13258     A. Al Malki   23       62         67                       Al Wehda   
13259     E. Guerrero   27       62         65                   CD Palestino   
13260         Hernáiz   20       62         69                    Albacete BP   
13261   H. Al Mansour   25       62         64                       Al Nassr   
13262         H. Paul   24       62         66               TSV 1860 München   
13263        S. Bauer   25       62         66       FC Admira Wacker Mödling   
13264      M. Chergui   29       62         62               Grenoble Foot 38   
13265      D. Gardner   28       62         62                Oldham Athletic   
13266    L. Bengtsson   20       62         73                    Hammarby IF   
13267    F. Jaramillo   22       62         70               Itagüí Leones FC   
13268      L. Garguła   37       62         62                  Miedź Legnica   
13269       S. Rivera   26       62         64            Jaguares de Córdoba   
13270        Vinicius   19       62         77                        Bologna   
13271    F. Sepúlveda   26       62         63                 CD Antofagasta   
13272       L. Spence   22       62         70                      Dundee FC   
13273      B. Lepistu   25       62         67                Kristiansund BK   
13274     A. Abruscia   27       62         62               TSV 1860 München   
13275     E. González   23       62         70                Boyacá Chicó FC   
13276      M. Al Amri   26       62         63                        Al Raed   
13277    J. Rebolledo   26       62         62               Deportes Iquique   
13278      C. Mamengi   17       62         79                     FC Utrecht   
13279    P. Mazzocchi   22       62         69                        Perugia   
13280       Y. Ammour   19       62         77                Montpellier HSC   
13281  Jwa Joon Hyeop   27       62         62                   Gyeongnam FC   
13282      O. Marrufo   25       62         65    Tiburones Rojos de Veracruz   
13283     Han Pengfei   25       62         66            Guizhou Hengfeng FC   

          Value    Wage  International Reputati

Se observa que al menos 48 filas tienen valores NaN para más de la mitad de las características de nuestro set de datos. Como 48 es equivalente a un porcentaje bastante pequeño de 18207 datos que hay, podemos eliminar estas filas problemáticas.

In [21]:
df = df.dropna(subset = ['International Reputation'])
df

Name  Age  Overall  Potential                 Club  \
0                L. Messi   31       94         94         FC Barcelona   
1       Cristiano Ronaldo   33       94         94             Juventus   
2               Neymar Jr   26       92         93  Paris Saint-Germain   
3                  De Gea   27       91         93    Manchester United   
4            K. De Bruyne   27       91         92      Manchester City   
...                   ...  ...      ...        ...                  ...   
18202        J. Lundstram   19       47         65      Crewe Alexandra   
18203  N. Christoffersson   19       47         63       Trelleborgs FF   
18204           B. Worman   16       47         67     Cambridge United   
18205      D. Walker-Rice   17       47         66      Tranmere Rovers   
18206           G. Nugent   16       46         66      Tranmere Rovers   

             Value      Wage  International Reputation  Weak Foot  \
0      110500000.0  565000.0                       5.0        4.0   
1       77000000.0  405000.0                       5.0        4.0   
2      118500000.0  290000.0                       5.0        5.0   
3       72000000.0  260000.0                       4.0        3.0   
4      102000000.0  355000.0                       4.0        5.0   
...            ...       ...                       ...        ...   
18202      60000.0    1000.0                       1.0        2.0   
18203      60000.0    1000.0                       1.0        2.0   
18204      60000.0    1000.0                       1.0        3.0   
18205      60000.0    1000.0                       1.0        3.0   
18206      60000.0    1000.0                       1.0        3.0   

       Skill Moves       Work Rate   Body Type Position Height Weight    LS  \
0              4.0  Medium/ Medium       Messi       RF    5.7    159  88+2   
1              5.0       High/ Low  C. Ronaldo       ST    6.2    183  91+3   
2              5.0    High/ Medium      Neymar       LW    5.9    150  84+3   
3              1.0  Medium/ Medium        Lean       GK    6.4    168   NaN   
4              4.0      High/ High      Normal      RCM   5.11    154  82+3   
...            ...             ...         ...      ...    ...    ...   ...   
18202          2.0  Medium/ Medium        Lean       CM    5.9    134  42+2   
18203          2.0  Medium/ Medium      Normal       ST    6.3    170  45+2   
18204          2.0  Medium/ Medium      Normal       ST    5.8    148  45+2   
18205          2.0  Medium/ Medium        Lean       RW   5.10    154  47+2   
18206          2.0  Medium/ Medium        Lean       CM   5.10    176  43+2   

         ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0      88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1      91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2      84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2   
18204  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2   
18205  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2   
18206  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2   

         CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0      84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1      81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2      81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   N

Luego, observamos qué columnas siguen afectadas por valores NaN.

In [22]:
df.columns[df.isna().any()].tolist()

['Club',
 'Position',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB']

Como resultado de la limpieza ahora tenemos un menos número de columnas que contienen valores NaN. A partir de la 4ta característica que tiene valores NaN podermos tratarlos calculando la media que de ese dato y usándola para sustituir el inválido.

Otra columna que nos da problemas por tener valores NaN es **Club**, pero como este atributo tiene peso en el valor de un jugador, no lo eliminaremos. En cambio, lo transformaremos de manera que podamos obtener valores numéricos sobre esta información para poder utilizarla en nuestro algoritmo.

La transformación consiste en añadir a la tabla el nombre de cada club y colocar un 1 si el jugador pertenece al club o un 0 en caso contrario.

In [23]:
df.loc[df.Club != df.Club]

Name  Age  Overall  Potential Club  Value  Wage  \
452        L. Paredes   24       80         85  NaN    0.0   0.0   
538      A. Granqvist   33       80         80  NaN    0.0   0.0   
568          A. Lunev   26       79         81  NaN    0.0   0.0   
677      I. Smolnikov   29       79         79  NaN    0.0   0.0   
874         A. Dzyuba   29       78         78  NaN    0.0   0.0   
...               ...  ...      ...        ...  ...    ...   ...   
17197        V. Kaith   21       55         64  NaN    0.0   0.0   
17215        K. Taimi   26       55         57  NaN    0.0   0.0   
17339        S. Passi   23       54         63  NaN    0.0   0.0   
17436  D. Lalhlimpuia   20       54         67  NaN    0.0   0.0   
17539        C. Singh   21       53         62  NaN    0.0   0.0   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
452                         2.0        4.0          4.0  Medium/ Medium   
538                         2.0        4.0          2.0    High/ Medium   
568                         1.0        3.0          1.0  Medium/ Medium   
677                         2.0        3.0          3.0      High/ High   
874                         2.0        3.0          3.0    High/ Medium   
...                         ...        ...          ...             ...   
17197                       1.0        2.0          1.0  Medium/ Medium   
17215                       1.0        3.0          2.0    Medium/ High   
17339                       1.0        3.0          2.0     Medium/ Low   
17436                       1.0        3.0          2.0  Medium/ Medium   
17539                       1.0        3.0          2.0    High/ Medium   

      Body Type Position Height Weight    LS    ST    RS    LW    LF    CF  \
452      Normal       CM   5.11    165  71+2  71+2  71+2  75+2  75+2  75+2   
538      Normal      LCB    6.4    185  62+2  62+2  62+2  56+2  58+2  58+2   
568      Normal       GK    6.2    176   NaN   NaN   NaN   NaN   NaN   NaN   
677        Lean       RB   5.10    154  70+2  70+2  70+2  73+2  72+2  72+2   
874      Stocky       ST    6.5    201  77+2  77+2  77+2  71+2  74+2  74+2   
...         ...      ...    ...    ...   ...   ...   ...   ...   ...   ...   
17197    Normal       GK    6.2    176   NaN   NaN   NaN   NaN   NaN   NaN   
17215    Normal       RB    6.4    187  46+2  46+2  46+2  45+2  44+2  44+2   
17339    Normal      NaN    5.9    143   NaN   NaN   NaN   NaN   NaN   NaN   
17436    Normal      NaN    6.0    168   NaN   NaN   NaN   NaN   NaN   NaN   
17539      Lean      NaN    6.3    174   NaN   NaN   NaN   NaN   NaN   NaN   

         RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
452    75+2  75+2  77+2  77+2  77+2  76+2  79+2  79+2  79+2  76+2  75+2  77+2   
538    58+2  56+2  58+2  58+2  58+2  57+2  64+2  64+2  64+2  57+2  68+2  74+2   
568     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677    72+2  73+2  73+2  73+2  73+2  75+2  74+2  74+2  74+2  75+2  78+2  75+2   
874    74+2  71+2  71+2  71+2  71+2  71+2  66+2  66+2  66+2  71+2  52+2  52+2   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17197   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17215  44+2  45+2  43+2  43+2  43+2  46+2  43+2  43+2  43+2  46+2  51+2  49+2   
17339   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17436   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17539   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
452    77+2  77+2  75+2  74+2  72+2  72+2  72+2  74+2      76.0       55.0   
538    74+2  74+2  68+2  70+2  79+2  79+2  79+2  70+2      49.0       51.0   
568     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      16.0       14.0   
677    75+2  75+2  78+2  78+2  73+2  73+2  73+2  78+2      73.0       61.0   

In [24]:
df = df.dropna(subset = ['Club'])

In [25]:
clb = df.pop("Club")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)
df

Name  Age  Overall  Potential        Value      Wage  \
0                L. Messi   31       94         94  110500000.0  565000.0   
1       Cristiano Ronaldo   33       94         94   77000000.0  405000.0   
2               Neymar Jr   26       92         93  118500000.0  290000.0   
3                  De Gea   27       91         93   72000000.0  260000.0   
4            K. De Bruyne   27       91         92  102000000.0  355000.0   
...                   ...  ...      ...        ...          ...       ...   
17913        J. Lundstram   19       47         65      60000.0    1000.0   
17914  N. Christoffersson   19       47         63      60000.0    1000.0   
17915           B. Worman   16       47         67      60000.0    1000.0   
17916      D. Walker-Rice   17       47         66      60000.0    1000.0   
17917           G. Nugent   16       46         66      60000.0    1000.0   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
0                           5.0        4.0          4.0  Medium/ Medium   
1                           5.0        4.0          5.0       High/ Low   
2                           5.0        5.0          5.0    High/ Medium   
3                           4.0        3.0          1.0  Medium/ Medium   
4                           4.0        5.0          4.0      High/ High   
...                         ...        ...          ...             ...   
17913                       1.0        2.0          2.0  Medium/ Medium   
17914                       1.0        2.0          2.0  Medium/ Medium   
17915                       1.0        3.0          2.0  Medium/ Medium   
17916                       1.0        3.0          2.0  Medium/ Medium   
17917                       1.0        3.0          2.0  Medium/ Medium   

        Body Type Position Height Weight    LS    ST    RS    LW    LF    CF  \
0           Messi       RF    5.7    159  88+2  88+2  88+2  92+2  93+2  93+2   
1      C. Ronaldo       ST    6.2    183  91+3  91+3  91+3  89+3  90+3  90+3   
2          Neymar       LW    5.9    150  84+3  84+3  84+3  89+3  89+3  89+3   
3            Lean       GK    6.4    168   NaN   NaN   NaN   NaN   NaN   NaN   
4          Normal      RCM   5.11    154  82+3  82+3  82+3  87+3  87+3  87+3   
...           ...      ...    ...    ...   ...   ...   ...   ...   ...   ...   
17913        Lean       CM    5.9    134  42+2  42+2  42+2  44+2  44+2  44+2   
17914      Normal       ST    6.3    170  45+2  45+2  45+2  39+2  42+2  42+2   
17915      Normal       ST    5.8    148  45+2  45+2  45+2  45+2  46+2  46+2   
17916        Lean       RW   5.10    154  47+2  47+2  47+2  47+2  46+2  46+2   
17917        Lean       CM   5.10    176  43+2  43+2  43+2  45+2  44+2  44+2   

         RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0      93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1      90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2      89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17913  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  44+2  44+2  45+2   
17914  42+2  39+2  40+2  40+2  40+2  38+2  35+2  35+2  35+2  38+2  30+2  31+2   
17915  46+2  45+2  44+2  44+2  44+2  44+2  38+2  38+2  38+2  44+2  34+2  30+2   
17916  46+2  47+2  45+2  45+2  45+2  46+2  39+2  39+2  39+2  46+2  36+2  32+2   
17917  44+2  45+2  45+2  45+2  45+2  46+2  45+2  45+2  45+2  46+2  46+2  46+2   

        CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0      61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1      61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2      60+3  60+3  65+3  60

Queremos hacer lo mismo con otras columnas que contienen valores no numéricos, pero que al relacionarlas con otras características nos dan información importante. Estas son: **Work Rate**, **Body Type** y **Position**.

In [26]:
df.loc[df["Work Rate"] != df["Work Rate"]]

Empty DataFrame
Columns: [Name, Age, Overall, Potential, Value, Wage, International Reputation, Weak Foot, Skill Moves, Work Rate, Body Type, Position, Height, Weight, LS, ST, RS, LW, LF, CF, RF, RW, LAM, CAM, RAM, LM, LCM, CM, RCM, RM, LWB, LDM, CDM, RDM, RWB, LB, LCB, CB, RCB, RB, Crossing, Finishing, HeadingAccuracy, ShortPassing, Volleys, Dribbling, Curve, FKAccuracy, LongPassing, BallControl, Acceleration, SprintSpeed, Agility, Reactions, Balance, ShotPower, Jumping, Stamina, Strength, LongShots, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Marking, StandingTackle, SlidingTackle, GKDiving, GKHandling, GKKicking, GKPositioning, GKReflexes, Release Clause, Nationality_Afghanistan, Nationality_Albania, Nationality_Algeria, Nationality_Andorra, Nationality_Angola, Nationality_Antigua & Barbuda, Nationality_Argentina, Nationality_Armenia, Nationality_Australia, Nationality_Austria, Nationality_Azerbaijan, Nationality_Barbados, Nationality_Belarus, Nationality_Belgium, Nationality_Belize, Nationality_Benin, Nationality_Bermuda, Nationality_Bolivia, Nationality_Bosnia Herzegovina, Nationality_Botswana, Nationality_Brazil, Nationality_Bulgaria, Nationality_Burkina Faso, Nationality_Burundi, Nationality_Cameroon, ...]
Index: []

In [27]:
df = df.dropna(subset = ['Work Rate'])

In [28]:
wrate = df.pop("Work Rate")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(wrate, prefix='w_rate').reset_index(drop=True)], axis=1, sort=False)
df.head()

Name  Age  Overall  Potential        Value      Wage  \
0           L. Messi   31       94         94  110500000.0  565000.0   
1  Cristiano Ronaldo   33       94         94   77000000.0  405000.0   
2          Neymar Jr   26       92         93  118500000.0  290000.0   
3             De Gea   27       91         93   72000000.0  260000.0   
4       K. De Bruyne   27       91         92  102000000.0  355000.0   

   International Reputation  Weak Foot  Skill Moves   Body Type Position  \
0                       5.0        4.0          4.0       Messi       RF   
1                       5.0        4.0          5.0  C. Ronaldo       ST   
2                       5.0        5.0          5.0      Neymar       LW   
3                       4.0        3.0          1.0        Lean       GK   
4                       4.0        5.0          4.0      Normal      RCM   

  Height Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0    5.7    159  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1    6.2    183  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2    5.9    150  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3    6.4    168   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   5.11    154  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   

    RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   

    LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0      75.0       91.0        76.0           61.0         87.0   

   Vision  Penalties  Composure  Marking  StandingTackle  SlidingTackle  \
0    94.0       75.0       96.0     33.0            28.0           26.0   
1    82.0       85.0       95.0     28.0            31.0           23.0   
2    87.0       81.0       94.0     27.0            24.0           33.0   
3    68.0       40.0       68.0     15.0  

In [29]:
btype = df.pop("Body Type")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(btype, prefix='b_type').reset_index(drop=True)], axis=1, sort=False)
df.head()

Name  Age  Overall  Potential        Value      Wage  \
0           L. Messi   31       94         94  110500000.0  565000.0   
1  Cristiano Ronaldo   33       94         94   77000000.0  405000.0   
2          Neymar Jr   26       92         93  118500000.0  290000.0   
3             De Gea   27       91         93   72000000.0  260000.0   
4       K. De Bruyne   27       91         92  102000000.0  355000.0   

   International Reputation  Weak Foot  Skill Moves Position Height Weight  \
0                       5.0        4.0          4.0       RF    5.7    159   
1                       5.0        4.0          5.0       ST    6.2    183   
2                       5.0        5.0          5.0       LW    5.9    150   
3                       4.0        3.0          1.0       GK    6.4    168   
4                       4.0        5.0          4.0      RCM   5.11    154   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN   NaN      17.0       13.0             21.0          50.0     13.0   
4  66+3  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0          

In [30]:
position = df.pop("Position")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(position, prefix='position').reset_index(drop=True)], axis=1, sort=False)
df.head()

Name  Age  Overall  Potential        Value      Wage  \
0           L. Messi   31       94         94  110500000.0  565000.0   
1  Cristiano Ronaldo   33       94         94   77000000.0  405000.0   
2          Neymar Jr   26       92         93  118500000.0  290000.0   
3             De Gea   27       91         93   72000000.0  260000.0   
4       K. De Bruyne   27       91         92  102000000.0  355000.0   

   International Reputation  Weak Foot  Skill Moves Height Weight    LS    ST  \
0                       5.0        4.0          4.0    5.7    159  88+2  88+2   
1                       5.0        4.0          5.0    6.2    183  91+3  91+3   
2                       5.0        5.0          5.0    5.9    150  84+3  84+3   
3                       4.0        3.0          1.0    6.4    168   NaN   NaN   
4                       4.0        5.0          4.0   5.11    154  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0    68.0       40.0   
4       91.0        76.0           61.0         87.0    94.0       79.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24.0           33.0       9.0         9.0   
3       68.0     15.0            21.0     

In [31]:
df.columns[df.isna().any()].tolist()

['LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB']

In [32]:
df.loc[:, 'LS':'RB']

LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0      88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1      91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2      84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17913  42+2  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2   
17914  45+2  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2   
17915  45+2  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2   
17916  47+2  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2   
17917  43+2  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2   

        LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0      84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1      81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2      81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17913  45+2  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2   
17914  35+2  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2  32+2  32+2   
17915  38+2  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2  28+2  28+2   
17916  39+2  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2  31+2  31+2   
17917  45+2  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2  47+2  47+2   

        RCB    RB  
0      47+2  59+2  
1      53+3  61+3  
2      47+3  60+3  
3       NaN   NaN  
4      66+3  73+3  
...     ...   ...  
17913  45+2  45+2  
17914  32+2  29+2  
17915  28+2  33+2  
17916  31+2  35+2  
17917  47+2  46+2  

[17918 rows x 26 columns]

In [33]:
list = df.loc[:, 'LS':'RB'].columns.tolist()

In [34]:
def add(x):
    """
    Add 
    
    """
    if x=="nan":
        return
    else:
        x = x.split('+')
        ret_val = int(x[0])+int(x[1])
        return ret_val

In [35]:
for i in range(len(list)):
    df[list[i]] = df[list[i]].astype(str)
    df[list[i]] = df[list[i]].apply(add)

In [36]:
df.loc[:, 'LS':'RB']

LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0      90.0  90.0  90.0  94.0  95.0  95.0  95.0  94.0  95.0  95.0  95.0  93.0   
1      94.0  94.0  94.0  92.0  93.0  93.0  93.0  92.0  91.0  91.0  91.0  91.0   
2      87.0  87.0  87.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  91.0   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      85.0  85.0  85.0  90.0  90.0  90.0  90.0  90.0  91.0  91.0  91.0  91.0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17913  44.0  44.0  44.0  46.0  46.0  46.0  46.0  46.0  47.0  47.0  47.0  46.0   
17914  47.0  47.0  47.0  41.0  44.0  44.0  44.0  41.0  42.0  42.0  42.0  40.0   
17915  47.0  47.0  47.0  47.0  48.0  48.0  48.0  47.0  46.0  46.0  46.0  46.0   
17916  49.0  49.0  49.0  49.0  48.0  48.0  48.0  49.0  47.0  47.0  47.0  48.0   
17917  45.0  45.0  45.0  47.0  46.0  46.0  46.0  47.0  47.0  47.0  47.0  48.0   

        LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0      86.0  86.0  86.0  93.0  66.0  63.0  63.0  63.0  66.0  61.0  49.0  49.0   
1      84.0  84.0  84.0  91.0  68.0  64.0  64.0  64.0  68.0  64.0  56.0  56.0   
2      84.0  84.0  84.0  91.0  68.0  63.0  63.0  63.0  68.0  63.0  50.0  50.0   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      90.0  90.0  90.0  91.0  80.0  80.0  80.0  80.0  80.0  76.0  69.0  69.0   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17913  47.0  47.0  47.0  46.0  46.0  47.0  47.0  47.0  46.0  47.0  47.0  47.0   
17914  37.0  37.0  37.0  40.0  32.0  33.0  33.0  33.0  32.0  31.0  34.0  34.0   
17915  40.0  40.0  40.0  46.0  36.0  32.0  32.0  32.0  36.0  35.0  30.0  30.0   
17916  41.0  41.0  41.0  48.0  38.0  34.0  34.0  34.0  38.0  37.0  33.0  33.0   
17917  47.0  47.0  47.0  48.0  48.0  48.0  48.0  48.0  48.0  48.0  49.0  49.0   

        RCB    RB  
0      49.0  61.0  
1      56.0  64.0  
2      50.0  63.0  
3       NaN   NaN  
4      69.0  76.0  
...     ...   ...  
17913  47.0  47.0  
17914  34.0  31.0  
17915  30.0  35.0  
17916  33.0  37.0  
17917  49.0  48.0  

[17918 rows x 26 columns]

Para cada una de estas columnas hallaremos la media, que es un valor razonable para sustituir los NaN que hay en cada una.

In [37]:
for i in range(len(list)):
    mean = df[list[i]].mean()
    df[list[i]] = df[list[i]].fillna(mean)

In [38]:
df

Name  Age  Overall  Potential        Value      Wage  \
0                L. Messi   31       94         94  110500000.0  565000.0   
1       Cristiano Ronaldo   33       94         94   77000000.0  405000.0   
2               Neymar Jr   26       92         93  118500000.0  290000.0   
3                  De Gea   27       91         93   72000000.0  260000.0   
4            K. De Bruyne   27       91         92  102000000.0  355000.0   
...                   ...  ...      ...        ...          ...       ...   
17913        J. Lundstram   19       47         65      60000.0    1000.0   
17914  N. Christoffersson   19       47         63      60000.0    1000.0   
17915           B. Worman   16       47         67      60000.0    1000.0   
17916      D. Walker-Rice   17       47         66      60000.0    1000.0   
17917           G. Nugent   16       46         66      60000.0    1000.0   

       International Reputation  Weak Foot  Skill Moves Height Weight  \
0                           5.0        4.0          4.0    5.7    159   
1                           5.0        4.0          5.0    6.2    183   
2                           5.0        5.0          5.0    5.9    150   
3                           4.0        3.0          1.0    6.4    168   
4                           4.0        5.0          4.0   5.11    154   
...                         ...        ...          ...    ...    ...   
17913                       1.0        2.0          2.0    5.9    134   
17914                       1.0        2.0          2.0    6.3    170   
17915                       1.0        3.0          2.0    5.8    148   
17916                       1.0        3.0          2.0   5.10    154   
17917                       1.0        3.0          2.0   5.10    176   

              LS         ST         RS         LW         LF         CF  \
0      90.000000  90.000000  90.000000  94.000000  95.000000  95.000000   
1      94.000000  94.000000  94.000000  92.000000  93.000000  93.000000   
2      87.000000  87.000000  87.000000  92.000000  92.000000  92.000000   
3      59.842647  59.842647  59.842647  61.066809  60.750471  60.750471   
4      85.000000  85.000000  85.000000  90.000000  90.000000  90.000000   
...          ...        ...        ...        ...        ...        ...   
17913  44.000000  44.000000  44.000000  46.000000  46.000000  46.000000   
17914  47.000000  47.000000  47.000000  41.000000  44.000000  44.000000   
17915  47.000000  47.000000  47.000000  47.000000  48.000000  48.000000   
17916  49.000000  49.000000  49.000000  49.000000  48.000000  48.000000   
17917  45.000000  45.000000  45.000000  47.000000  46.000000  46.000000   

              RF         RW        LAM        CAM        RAM         LM  \
0      95.000000  94.000000  95.000000  95.000000  95.000000  93.000000   
1      93.000000  92.000000  91.000000  91.000000  91.000000  91.000000   
2      92.000000  92.000000  92.000000  92.000000  92.000000  91.000000   
3      60.750471  61.066809  61.015195  61.015195  61.015195  61.760455   
4      90.000000  90.000000  91.000000  91.000000  91.000000  91.000000   
...          ...        ...        ...        ...        ...        ...   
17913  46.000000  46.000000  47.000000  47.000000  47.000000  46.000000   
17914  44.000000  41.000000  42.000000  42.000000  42.000000  40.000000   
17915  48.000000  47.000000  46.000000  46.000000  46.000000  46.000000   
17916  48.000000  49.000000  47.000000  47.000000  47.000000  48.000000   
17917  46.000000  47.000000  47.000000  47.000000  47.000000  48.000000   

             LCM         CM        RCM         RM        LWB        LDM  \
0      86.000000  86.000000  86.000000  93.000000  66.000000  63.000000   
1      84.000000  84.000000  84.000000  91.000000  68.000000  64.000000   
2      84.000000  84.000000  84.000000  91.000000  68.000000  63.000000   
3      60.244569  60.244569  60.244569  61.760455  59.516891  58.831659   
4      90.000000  90.000000  90.000000  91

Finalmente, ya nos hemos deshecho de los valores NaN presentes en nuestro set de datos.
Sin embargo, hasta el momento habíamos mantenido la columna **Name** por tratarse de un valor informativo y de referencia a lo largo del análisis y tratamiento de los datos, pero por tratarse de un dato no numérico, será eliminado del set.

In [39]:
df.pop('Name')

0                  L. Messi
1         Cristiano Ronaldo
2                 Neymar Jr
3                    De Gea
4              K. De Bruyne
                ...        
17913          J. Lundstram
17914    N. Christoffersson
17915             B. Worman
17916        D. Walker-Rice
17917             G. Nugent
Name: Name, Length: 17918, dtype: object

In [41]:
val = df["Value"]

## Predicción

En este punto, ya los datos están limpios y en el formato que nuestro algoritmo acepta. El siguiente paso es entrenar y predecir, pero antes debemos dividir el set de datos en: entrenamiento y test. La proporción será 80% y 20%, respectivamente.

In [42]:
X_train, X_test, y_train, y_test = train_test_split(df,df["Value"], test_size=0.20, random_state=50)

Ahora que ya el set está dividido, procedemos a normalizar los datos.

In [43]:
norm = preprocessing.MinMaxScaler()
X_train_norm = norm.fit_transform(X_train)

In [44]:
X_test_norm = norm.transform(X_test)

Finalmente, hacemos la **regresión lineal** para obtener la métrica $R^2$

In [45]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

In [46]:
preds = reg.predict(X_test)

In [47]:
mean_squared_error(y_test, preds)

3.4760104287989156e-17

In [48]:
r2_score(preds, y_test)

1.0

Sabemos que para las regresiones que si la métrica obtenida es cercana o igual a 1, se trata de un ajuste bastante bueno y, por lo tanto, las predicciones que se hagan con ese modelo serán bastante correctas.

Así, al haber obtenido una métrica igual a 1, concluimos que las predicciones que se hacen este algoritmo sobre el valor monetario de un jugador de fútbol son fiables.